document how `--gc:arc` and `--gc:orc` work (#555)
nim-works/nimskull: doc/gc.rst

types with lifetime-tracking hooks.

ORC
===

The problem with plain reference counting is that it is not able to handle
reference cycles. A reference cycle exists when a cell either directly or
indirectly stores a reference (i.e. `ref`) to itself. Example:

.. code-block:: nim
  type A = object
    x: ref A

  var a = A(x: nil) # okay; no cycle exists and the refcounter is 1
  a.x = a # a cycle is introduced! the refcounter is 2

  # once `a` gets destroyed, the refcounter becomes 1, but it's not possible
  # for it to reach 0 from there
  a.x = nil # explictly breaking the cycle would work

  type
    B = object
      c: ref C
    C = object
      b: ref B

  var b = B()
  b.c = C() # okay; no cycle exists
  b.c.b = b # an indirect cycle is introduced!

This is where ORC becomes relevant. ORC is automatic reference-counting with a
run-time cycle collector (the letter 'O' in ORC is meant to represent a cycle).

For easier visulization, it makes sense to view a managed heap cell as a *node*,
and a `ref` as an *edge* in a directed cyclic graph. The cycle collector is
responsible for freeing cells only referenced as part of references cycles.

> The algorithm

Reviewer: ```suggestion
The Algorithm
```

-------------

> The cycle colletor used for ORC is based on "trial deletion". As the first

Reviewer: ```suggestion
The cycle collector used for ORC is based on "trial deletion". As the first
```

step, all edges reachable from potential *cycle roots* are temporarily
deleted. The second step restores the outgoing edges for nodes (and nodes
reachable from them) that still have incoming edges after the temporary
deletion. Once done, all nodes that have no incoming edges are known to only
be alive because of a reference cycle, and can now be freed.

> The implementation

Reviewer: ```suggestion
The Implementation
```

------------------

> 

Reviewer: ```suggestion

Collection is implemented with a 2-pass graph colouring algorithm for detection,
and a 2-step memory collection process.

```

This intro might help setup the mental model described in the paragraphs that follow.

Author: Yep, that's better. Thanks.

Each cell is assigned a *color*:
- *black*: the cell is alive (new cells start with this color)
- *gray*: the cell was visited by the collector already
- *white*: the cell is dead and can be cleaned up

The cycle collector only considers *edges* and *nodes* through which a cycle
is possible, the others are ignored and never touched.

During the first pass, the sub-graph of each cell in the input set is traversed

> and each reachable *black* node (i.e. cell) as *gray* and temporarily deleting

Reviewer: ```suggestion
and each reachable *black* node (i.e. cell) set to *gray* and temporarily deleting
```

> its edges (i.e. decrementing the refcount of the referenced cell by one).

Reviewer: ```suggestion
its edges (i.e. decrementing the refcount of each referenced cell by one).
```

*Gray* nodes are not traversed further -- they've already been visited.

The second pass traverses the sub-graph of each cell in the input set and
forward-propagates the color *black* from all nodes with the color *gray* and
a refcount > 0 -- they are being kept alive by something not part of a cycle.
The others are marked as *white*.

.. note:: the second pass is skipped if all cells marked as *gray* have no
  incoming edges left

Finally, all cells part of the input sub-graphs with either *white* or *gray*
(depending on whether the second pass was skipped) as the color are cleaned
up. Clean up happens in two steps: first, all outgoing edges relevant to the
cycle collector are physically removed (by setting the `ref` values to `nil`).
Then, the cell is disposed by calling its destructor (if one exists) an

> freeing the underlying memory location.

Reviewer: ```suggestion
Finally, all cells of the input sub-graph that are either *white* or *gray* are
cleaned up in a two step process: first, all outgoing edges relevant to the
cycle collector are physically removed (by setting the `ref` values to `nil`).
Then, the cell is disposed by calling its destructor (if one exists) an freeing
the underlying memory location.
```

Not sure if that's still accurate, but it reads a bit better. I dropped the clarification for the gray cells as the note and paragraph prior cover that.

Author: I think it's mostly still accurate, although only mentioning a single sub-graph might be a possible source of confusion. The original paragraph was not entirely correct, however, so I rewrote it as a whole.

> Input set

Reviewer: ```suggestion
Input Set
```

~~~~~~~~~

Which cells the input set contains depends on how the cycle collection was
invoked. If it's a full collection, the input set contains all potential
*cycle roots*, but for a partial collection (i.e. started by calling
`GC_partialCollect`), only the roots with an index greater than or equal to
the speicifed `limit` are considered.

> Potential cycle roots

Reviewer: ```suggestion
Potential Cycle Roots
```

~~~~~~~~~~~~~~~~~~~~~

A central part of ORC is registering potential *cycle roots*. If static
analysis of a `ref`'s type yields that no cycles are possible through it,
it is never treated as an edge by the collector.

When an edge through which a reference cycle can happen is removed (i.e. a
`ref` is destroyed), the cell is remembered as a potential *cycle root*. Doing

> this on edge removal instad of creation has the benefit that sub-graphs that

Reviewer: ```suggestion
this on edge removal instead of creation has the benefit that sub-graphs that
```

are definitely kept alive from outside a cycle are not already scanned during
the first two passes.

Once the list of potential *cycle roots* reaches a certain threshold, a full
cycle collection is immediately run (if cycle collection is not disabled at
program run-time). Since roots are only registered when a `ref` value (that
can form a reference cycle) is destroyed, cycle collection can only happen
when copying, sinking, or destroying a `ref`.

The threshold is dynamic (but can be made static via the `nimFixedOrc` define).
If more than 50% of the visited cells were freed during a *full* collection,
the threshold is reset to the default value -- otherwise it's increased until
an implementation defined upper bound is reached.

> Gathering edges

Reviewer: ```suggestion
Gathering Edges
```

~~~~~~~~~~~~~~~

> To know about the outgoing edges of a cell, the cell's attached `=trace` hooks is

Reviewer: ```suggestion
To know about the outgoing edges of a cell, the cell's attached `=trace` hook is
```

invoked. It is responsible for collecting all directly reachable relevant
`ref`s of the cell to a list provided by the collector.

> For more information, see the `documentation <destructors.html#lifetimeminustracking-hooks-nimeqtrace-hook>`_
> of the `=trace` hook.

Reviewer: ```suggestion
For more information, see the `=trace` hook `documentation 
<destructors.html#lifetimeminustracking-hooks-nimeqtrace-hook>`_.
```

> Thread safety

Reviewer: ```suggestion
Thread Safety
```

The `NimScript <nims.html>`_ target uses the memory management strategy built into
the Nim compiler.

ARC
===

ARC, which is short for automatic reference counting, is a memory-management
strategy based on plain reference counting. Each managed heap cell (that is,
a location allocated via either `new` or by using the object construction syntax
where the type is a `ref T`) also stores a reference counter.

> When a managed cell is allocated, its counter starts at `1` (because the
> allocation created a single `ref` handle). Creating a `ref` to the cell
> increments the counter, and destroying one decrements it It is incremented when
> a new `ref` to the cell is created, and decremented when a `ref` to it is
> destroyed.

Reviewer: ```suggestion
When a managed cell is allocated, its counter starts at `1` (because the
allocation created a single `ref` handle). Creating a `ref` to the cell
increments the counter, and destroying one decrements it.
```
